features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/metric/SessionEndedMetricDispatcher.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.rum.internal.metric
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.rum.internal.domain.scope.RumViewManagerScope
+import com.datadog.android.rum.model.ViewEvent
+
+/**
+ * Implementation of [SessionMetricDispatcher].
+ */
+internal class SessionEndedMetricDispatcher(private val internalLogger: InternalLogger) :
+    SessionMetricDispatcher {
+
+    private val metricsMap by lazy { linkedMapOf<String, SessionEndedMetric>() }
+
+    override fun startMetric(
+        sessionId: String,
+        precondition: String
+    ) {
+        metricsMap[sessionId] = SessionEndedMetric(sessionId, precondition)
+    }
+
+    override fun endMetric(sessionId: String) {
+        val metric = metricsMap[sessionId]
+
+        metric?.let {
+            internalLogger.logMetric(
+                messageBuilder = { SessionEndedMetric.RUM_SESSION_ENDED_METRIC_NAME },
+                additionalProperties = metric.toMetricAttributes()
+            )
+        }
+
+        metricsMap.remove(sessionId)
+    }
+
+    override fun onSessionStopped(sessionId: String) {
+        metricsMap[sessionId]?.onSessionStopped()
+    }
+
+    override fun onViewTracked(sessionId: String, viewEvent: ViewEvent) {
+        val result = metricsMap[sessionId]?.onViewTracked(viewEvent) ?: false
+        if (result.not()) {
+            internalLogger.log(
+                InternalLogger.Level.INFO,
+                InternalLogger.Target.MAINTAINER,
+                { buildViewTrackError(sessionId, viewEvent) }
+            )
+        }
+    }
+
+    override fun onSdkErrorTracked(errorKind: String?) {
```

Comment:
what is the supposed call site for this? If we can supply `sessionId` as an argument here, then we won't need to have `LinkedHashMap` and do `last` (not a big deal though).

Reply:
I plan to call it inside `TelemetryEventHandler` when handling a error, I am not sure I can have the session id there

Reply:
You can get it there in a few ways (depends on the exact place you want to use):

* if it is inside `sdkCore.getFeature(Feature.RUM_FEATURE_NAME)?.withWriteContext` you can get it from `DatadogContext`
* you also have `onSessionStarted` there, so you can just keep the current session ID in the property

Reply:
OK, then I added `sessionId` as parameter inside this function